M2/Macaulay2/m2/gb.m2

```diff
@@ -22,463 +29,478 @@ RawStatusCodes = new HashTable from {
 
 toEngineNat  := n -> if n === infinity then -1 else n
 
+---------------------------
+
+-- also used in res.m2 and tests/engine/raw-gb.m2
+isComputationDone = c -> if RawStatusCodes#?c then "done" === RawStatusCodes#c else error "unknown computation status code"
+
+processStrategy := strategies -> sum(flatten {strategies}, strategy ->
+    if RawStrategyCodes#?strategy then RawStrategyCodes#strategy else error "gb: unknown strategy encountered")
+
+-- TODO:
+--   move the codes below into a hashtable above
+--   implement condition checking mechanism similar to Colon.m2
+--   F4 is still used by groebnerBasis, is it the same as LinearAlgebra?
+--   integrate FGLM, GroebnerWalk, ThreadedGB, Hilbert driven Buchberger, etc.
+processAlgorithm := (alg, m) -> (
+    R := ring m;
+    k := ultimate(coefficientRing, R);
+    -- compatibility checking
+    if (alg === Homogeneous or alg === Homogeneous2) and not isHomogeneous m
+    then error "gb: homogeneous algorithm specified with inhomogeneous matrrix";
+    if k === ZZ and alg =!= Inhomogeneous
+    then error "gb: only the algorithm 'Inhomogeneous' may be used with base ring ZZ";
+    if R.?FlatMonoid and not R.FlatMonoid.Options.Global and alg =!= Inhomogeneous
+    then error "gb: only the algorithm 'Inhomogeneous' may be used with a non-global monomial ordering";
+    -- return algorithm code
+    -- Keep these in sync with the values in e/comp-gb.cpp (TODO: where?)
+    if      alg === Homogeneous   then 1
+    else if alg === Inhomogeneous then 2
+    -- else if alg === F4            then error "the F4 algorithm option has been replaced by LinearAlgebra"
+    -- else if alg === Faugere       then error "the Faugere algorithm option has been replaced by LinearAlgebra"
+    else if alg === Sugarless     then 4
+    else if alg === Homogeneous2  then 5
+    else if alg === LinearAlgebra then (warnexp(); 6)
+    else if alg === Toric         then (warnexp(); 7)
+    else if alg === Test          then 8
+    else error "unknown algorithm encountered")
+
+-----------------------------------------------------------------------------
+-- GroebnerBasis type declarations and basic constructors and functions
+-----------------------------------------------------------------------------
+
+GroebnerBasis = new Type of MutableHashTable
+GroebnerBasis.synonym = "Gröbner basis"
+
 GroebnerBasisOptions = new Type of OptionTable
+GroebnerBasisOptions.synonym = "Gröbner basis options"
+
+-- m:    a Matrix
+-- type: an GroebnerBasisOptions
+-- opts: an OptionTable
+-- TOOD: document this
+new GroebnerBasis from Sequence := (GB, S) -> (
+    (m, type, opts) := if #S === 3 then S else error "GroebnerBasis: expected three initial values";
+    -- TODO: implement recursive type checking
+    if not instance(m, Matrix) or not instance(type, GroebnerBasisOptions) or not instance(opts, OptionTable)
+    then error "GroebnerBasis: expected a sequence of type (Matrix, GroebnerBasisOptions, OptionTable)";
```

Comment:
I agree that it's inconvenient to check the types, but "newgb" is also an inconvenient way to define a constructor. It would be better to allow something like this:
```m2
new GroebnerBasis from (Matrix, GroebnerBasisOptions, OptionTable) := ...
```
What do you think?

Reply:
I think that would be hard to implement in 1 day.

Reply:
Although, it might be nice to have!

Reply:
I'd rather postpone this to when we get to unifying `gb` and `groebnerBasis` methods, which should hopefully be soon.